README.md
# Maker Assist

## Introduction
An assistant for Makers and Developers.

## Changes
All notable changes to this project will be documented in this [Change log](CHANGELOG.md).

## Requirements
For installing and running, you will need Ansible (>= 2.7), Python (>=3.6.2) and PIP (>= 1.5).

## Installation

### Cloning
``` bash
git clone --recurse-submodules https://github.com/marcio-pessoa/makerassist.git
```

### Install
Change to work directory and run Ansible Playbook:

``` bash
cd makerassist
ansible-playbook ma.yaml
```

Running:
``` bash
./ma
```

For help:
``` bash
./ma -h
```

---


CHANGELOG.md
# Maker Assist

## Change log
All notable changes to this project will be documented in this file.

### [0.9] - 2019-12-30
#### Fixed
- Command verify not reporting error for unknown devices.

### [0.89] - 2019-12-22
#### Fixed
- Command list output when runing on verbosity Critical.

### [0.88] - 2019-09-04
#### Removed
- Python 2 support.
#### Added
- Logging package messages support.

### [0.78] - 2019-07-08
#### Fixed
- Verboseless messages.

### [0.77] - 2019-07-07
#### Added
- Ansible installer.

### [0.76] - 2019-03-04
#### Added
- MicroPython support.

### [0.73] - 2019-01-11
#### Fixed
- Command list command description.

### [0.72] - 2019-01-08
#### Fixed
- Serial intermittent disconnection.

### [0.7] - 2019-01-01
#### Added
- Python 3 ready!

### [0.6] - 2018-11-16
#### Changed
- Subprocess improved performance

### [0.51] - 2018-02-11
#### Added
- List only enabled devices.
#### Changed
- List device link status on debug option 4 (no more 3).
#### Changed
- Supressed copyright messages from --version option.

### [0.42] - 2017-04-30
#### Added
- list command can now show only connected devices.

### [0.42] - 2017-03-06
#### Added
- list command can force check on a specific interface.

### [0.41] - 2017-02-27
#### Fixed
- Auto detection was overlapping explicit device ID definition.
- Command list performance issues

### [0.4] - 2017-02-24
#### Added
- Device auto detection feature.

### [0.33] - 2017-02-21
#### Changed
- Some changes to support future migration to Python 3.

### [0.3] - 2017-02-15
#### Added
- Argparse support.
#### Fixed
- Help messages and user input arguments and options.

### [0.25] - 2017-02-12
#### Added
- Connection status is available to list command.

### [0.24] - 2017-02-07
#### Changed
- Help messages optimization.

### [0.23] - 2016-04-12
#### Changed
- Code optionazation, separated into some package files.

### [0.22] - 2015-10-18
#### Fixed
- list option fixed.

### [0.21] - 2015-09-11
#### Added
- Availability check for libraries check before import.

### [0.2] - 2015-08-19
#### Changed
- JSON device file optimization and organization.

### [0.1] - 2015-07-17
#### Changed
- Converted original Bash code to Python. All features ported.

### [0.02] - 2014-02-07
#### Fixed
- Minor bug fix.

### [0.01] - 2014-02-06
#### Added
- First version written in Bash.


tools/session.py
"""
---
name: session.py
description: Device session package
copyright: 2016-2019 Marcio Pessoa
people:
  developers:
  - name: Marcio Pessoa
    email: marcio.pessoa@gmail.com
change-log:
  2019-09-03
  - version: 0.7
    fixed: String receive and parsing (only on Python 3).
  2019-01-08
  - version: 0.06
    fixed: Minor updated to support serial on Python 3.
  2017-07-27
  - version: 0.05b
    added: Featured colours to 'ok' and 'nok' received strings.
  2017-06-04
  - version: 0.04b
    added: Added comment parser to '()'
  2017-06-02
  - version: 0.03b
    added: Added comment parser to ';'
  2017-05-11
  - version: 0.02b
    added: Added is_ready() method.
    fixed: Added new line to send and receive screen output.
    added: Added command string filter in send() method.
  2017-02-21
  - version: 0.01b
    added: Added information messages.
  2016-02-19
  - version: 0.00b
    added: Scrach version.
"""

import sys
import os.path
import os
import re
from time import sleep
from socket import gethostbyname
import serial


class Session:  # pylint: disable=too-many-instance-attributes,too-many-public-methods
    """
    description:
    """

    __version__ = 0.7

    def __init__(self, data):
        self.interface = None
        self.data = None
        self.gcode = None
        self.comm_serial_path = ''
        self.comm_serial_speed = 0
        self.comm_serial_terminal_echo = True
        self.comm_serial_terminal_end_of_line = 'CRLF'
        self.comm_serial_delay = 0
        self.comm_network_address = None
        self.retry = 3  # times
        self.timeout = 100  # milliseconds
        self.session = None
        self.reset()
        self.load(data)

    def reset(self):
        """
        description:
        """
        self.data = None
        self.comm_serial_path = ''
        self.comm_serial_speed = 0
        self.comm_serial_terminal_echo = True
        self.comm_serial_terminal_end_of_line = 'CRLF'
        self.comm_serial_delay = 0
        self.comm_network_address = None
        self.retry = 3  # times
        self.timeout = 100  # milliseconds

    def load(self, data):
        """
        description:
        """
        if data == []:
            return None
        self.data = data
        # Optional keys
        try:
            self.comm_serial_path = \
                self.data["serial"].get('path', self.comm_serial_path)
            self.comm_serial_speed = \
                self.data["serial"].get('speed', self.comm_serial_speed)
            self.comm_serial_delay = \
                self.data["serial"].get('delay', self.comm_serial_delay)
            self.comm_serial_terminal_echo = \
                self.data["serial"].get('terminal_echo', self.comm_serial_terminal_echo)
            self.comm_serial_terminal_end_of_line = \
                self.data["serial"].get('terminal_end_of_line',
                                        self.comm_serial_terminal_end_of_line)
        except KeyError:
            pass
        # Network configuration
        try:
            self.comm_network_address = (
                self.data["network"].get('address', self.comm_network_address))
            # Get IP address from host name
            self.comm_network_address = gethostbyname(self.comm_network_address)
        except BaseException:
            pass
        return 0

    def is_connected(self):
        """
        description:
        """
        if self.is_connected_serial():
            return True
        return False

    def is_connected_serial(self):
        """
        description:
        """
        return os.path.exists(self.comm_serial_path)

    def stop(self):
        """
        description:
        """
        if self.session:
            self.session.close()
        self.reset()

    def start(self):
        """
        description:
        """
        if not self.data:
            return None
        # echo.infoln("Connecting...", 1)
        if self.is_connected_serial():
            try:
                self.session = serial.Serial()
                self.session.port = self.comm_serial_path
                self.session.baudrate = self.comm_serial_speed
                self.session.timeout = 1
                self.session.open()
            except BaseException:
                # echo.warnln('Operation not completed.', 2)
                return False
            sleep(self.comm_serial_delay / 1000)
            # echo.infoln('Speed: ' + str(self.comm_serial_speed) + ' bps', 2)
            return self.session
        # echo.warnln('Device is not connected.', 2)
        return True

    def info(self):
        """
        description:
        """
        # echo.infoln('Session...')
        if not self.data:
            return
        # if self.comm_serial_path is not None:
            # echo.infoln('Startup delay: ' + str(self.comm_serial_delay) + ' ms', 1)
            # echo.infoln('Serial: ' + str(self.comm_serial_path), 1)
        # if self.comm_network_address is not None:
            # echo.infoln('Network: ' + str(self.comm_network_address), 1)

    def run(self):
        """
        description:
        """
        # echo.infoln('Running program...')
        go_on = False
        while not go_on:
            go_on = self.is_ready()
        lines = len(self.gcode)
        counter = 0
        while True:
            if go_on:
                if counter == lines:
                    break
                line = self.gcode[counter]
                counter += 1
                if self.send(line):
                    continue
            received = self.receive()
            go_on = bool(re.search('ok', str(received)))

    def is_ready(self):
        """
        description:
        """
        i = 0
        while True:
            content = self.receive()
            if content is False:
                i += 1
            else:
                i = 0
            if i > 1:
                break
        return True

    def play(self):  # pylint: disable=no-self-use
        """
        description:
        """
        return False

    def pause(self):  # pylint: disable=no-self-use
        """
        description:
        """
        return False

    def send_expect(self, command, expected, timeout=0, retry=0):
        """Send a command and receive a message.

        Retrieves rows pertaining to the given keys from the Table
        instance represented by big_table.  Silly things may happen if
        other_silly_variable is not None.

        Args:
            command:
            expected:
                A list with three items:
                [0] Header
                    - Description: Field used to start a message.
                    - Default value: This is a required field.
                [1] Payload
                    - Description: Message content.
                                   Also kown as data field.
                    - Default value: This is a required field.
                [2] Trailer
                    - Description: Control field used to identify end of
                                   message.
                                   Also known as footer field.
                    - Default value: '\n'
            timeout:
            retry:

        Returns:
            A dict mapping keys to the corresponding table row data
            fetched. Each row is represented as a tuple of strings. For
            example:

            {'Serak': ('Rigel VII', 'Preparer'),
             'Zim': ('Irk', 'Invader'),
             'Lrrr': ('Omicron Persei 8', 'Emperor')}

            If a key from the keys argument is missing from the dictionary,
            then that row was not found in the table.

        Raises:
            IOError: An error occurred accessing the bigtable.Table object.
        """
        if retry == 0:
            retry = self.retry
        if timeout == 0:
            timeout = self.timeout
        try:
            self.send(command)
        except IOError:
            return True
        serial_line = ""
        for _ in range(retry):
            sleep(int((timeout) / 1000))
            serial_line += self.receive()
            if expected in serial_line:
                return serial_line
        # echo.erro("Command return lost.")
        return True

    def receive(self):
        """
        description: Just receive a message
        """
        try:
            received = self.session.readline().rstrip()
            received = str(received, 'utf-8')
        except IOError:
            return True
        if received == "":  # or received == "\n" or received == "\r":
            return False
        # Change color based on device response ('ok' or 'nok')
        if re.search('nok', str(received)):
            # echo.codeln(received, 'red', attrs=['bold'])
            return received
        if re.search('ok', str(received)):
            # echo.codeln(received, 'green', attrs=['bold'])
            return received
        return received

    def send(self, command):
        """
        description: Just send a message
        """
        # Store comments
        try:
            comment = command[re.search(';', command).span()[0]:]
        except BaseException:
            comment = ''
        if comment == '':
            try:
                comment = command[re.search('\(', command).span()[0]:]  # pylint: disable=anomalous-backslash-in-string
            except BaseException:
                comment = ''
        # Remove comments
        command = re.sub(r'(?:_a)?\(([^(]*)$', '\n', command)
        command = re.sub(r'(?:_a)?\;([^;]*)$', '\n', command)
        # Trim start and end spaces
        command = command.strip(' ').rstrip()
        comment = comment.strip(' ').rstrip()
        # Ignore blank lines
        if command == '':
            # if comment != '':
                # echo.codeln(comment)
            return True
        # echo.code(command, attrs=['bold'])
        # echo.codeln('  ' + comment)
        try:
            self.session.write((command + '\n').encode())
        except IOError:
            return True
        return False

    def send_wait(self, command):
        """
        description:
        """
        self.send(command)
        while True:
            received = self.receive()
            if re.search('ok', str(received)):
                break

    def set_retry(self, retry):
        """
        description: Set default number of retries
        """
        self.retry = retry

    def set_timeout(self, timeout):
        """
        description: Set default timeout value
        """
        self.timeout = timeout

    def clear(self):
        """
        description: Clear message buffer
        """
        while self.receive():
            continue

    def set_program(self, data):
        """
        description: Set G-Code
        """
        if not data or data == '':
            # echo.erroln('Invalid G-code file.')
            sys.exit(True)
        self.gcode = data


tools/device.py
"""
---
name: device.py
description: Device package
copyright: 2016-2019 Marcio Pessoa
people:
  developers:
  - name: Marcio Pessoa
    email: marcio.pessoa@gmail.com
change-log:
  2019-09-07
  - version: 0.13
    fixed: Adequation to pylint.
  2019-02-12
  - version: 0.12
    fixed: Auto detection was returning always the last device.
  2019-01-25
  - version: 0.11b
    fixed: Auto detection was returning invalid device name.
  2018-04-02
  - version: 0.10b
    changed: Only configured interfaces are shown.
  2018-02-11
  - version: 0.09b
    added: is_enable method.
  2017-06-06
  - version: 0.08b
    added: Terminal line feed and carriage return customization.
    added: Terminal local echo customization.
  2017-06-06
  - version: 0.07b
    changed: Removed Warning messages.
  2017-05-11
  - version: 0.06b
    added: Added startup() method.
  2017-05-08
  - version: 0.05b
    added: Added presentation() method.
  2017-04-01
  - version: 0.04b
    changed: Added version information.
  2017-02-22
  - version: 0.03b
    changed: Removed method DeviceProperties.items_extended().
    added: Method DeviceProperties.is_network_connected().
  2017-02-21
  - version: 0.02b
    added: Information messages.
    added: Verbose error messages.
  2017-02-20
  - version: 0.01b
    added: Changed except ',' to 'as' to addopt Python 3 style.
  2016-05-12
  - version: 0.00b
    added: first version.
"""

import sys
from tools.session import Session


class DeviceProperties:  # pylint: disable=too-many-instance-attributes
    """
    description:
    """

    __version__ = 0.13

    def __init__(self, data):
        self.version = 0.13
        self.__id = None
        self.interface = None
        self.system_plat = None
        self.system_mark = None
        self.system_desc = None
        self.system_arch = None
        self.system_path = None
        self.system_work = None
        self.system_logs = None
        self.load(data)
        self.reset()

    def load(self, data):
        """
        description:
        """
        self.data = data

    def is_enable(self):
        """
        description:
        """
        enable = False
        try:
            enable = self.data["device"][self.__id].get('enable', True)
        except BaseException:
            pass
        return enable

    def get_id(self):
        """
        description:
        """
        return self.__id

    def set(self, device_id):
        """Set a device to be used.

        Used to set a device and import all elements.

        Attributes:
            element: An element name (like x1, x2, x6, etc.).
        """
        if not device_id:
            return
        self.__id = device_id
        # Is device present in configuration file?
        # try:
            # echo.check_id = self.data["device"][self.__id]
        # except KeyError:
            # echo.erroln('Device is not present in configuration file.')
            # sys.exit(True)
        # Check mandatory keys.
        try:
            self.system_plat = self.data["device"][self.__id]["system"]["plat"]
            self.system_mark = self.data["device"][self.__id]["system"]["mark"]
            self.system_desc = self.data["device"][self.__id]["system"]["desc"]
            self.system_arch = self.data["device"][self.__id]["system"]["arch"]
            self.system_path = self.data["device"][self.__id]["system"]['path']
            self.system_work = self.data["device"][self.__id]["system"]['work']
            self.system_logs = self.data["device"][self.__id]["system"]['logs']
        except KeyError: # as err:
            # echo.erroln('Mandatory key is absent: %s' % (err))
            sys.exit(True)

    def list(self):
        """Fetches device IDs from a dictionary.

        Args:
            None

        Returns:
            A list of device IDs. For example:

            ['x1', 'x2', 'x3', 'x4']

        Raises:
            None
        """
        elements = []
        for i in self.data["device"]:
            elements.append(i)
        elements.sort()
        return elements

    def get(self):
        """
        description:
        """
        try:
            return self.data["device"][self.__id]
        except BaseException:
            return []

    def get_comm(self):
        """
        description:
        """
        try:
            return self.data["device"][self.__id]["comm"]
        except BaseException:
            return []

    def get_objects(self):
        """
        description:
        """
        try:
            return self.data["device"][self.__id]["object"]
        except BaseException:
            return []

    def get_endup(self):
        """
        description:
        """
        try:
            return self.data["device"][self.__id]["endup"]
        except BaseException:
            return []

    def get_startup(self):
        """
        description:
        """
        try:
            return self.data["device"][self.__id]["startup"]
        except BaseException:
            return []

    def get_control(self):
        """
        description:
        """
        try:
            return self.data["device"][self.__id]["control"]
        except BaseException:
            return []

    def reset(self):
        """Reset device default properties"""
        self.__id = None
        self.system_plat = ""
        self.system_mark = ""
        self.system_desc = ""
        self.system_arch = ""
        self.system_path = ""
        self.system_logs = ""

    def set_interface(self, interface):
        """
        description:
        """
        self.interface = interface

    def info(self):
        """
        description:
        """
        if not self.__id:
            return None
        return \
            '    ID: ' + str(self.__id) + '\n' + \
            '    Name: ' + str(self.system_plat) + ' ' + \
            'Mark ' + str(self.system_mark) + '\n' + \
            '    Description: ' + str(self.system_desc)

    def detect(self):
        """
        description:
        """
        self.reset()
        ids = []
        for device_id in self.list():
            self.set(device_id)
            session = Session(self.get_comm())
            if not self.is_enable():
                continue
            if session.is_connected_serial():
                ids.append(str(device_id).encode("utf-8"))
            else:
                session.reset()
            self.reset()
        devices = len(ids)
        if devices == 1:
            self.set(ids[0].decode('utf-8'))
        elif devices < 1:
            self.set(None)
        elif devices > 1:
            self.set(False)
        return ids
